### PR TITLE
Fix logsumexp op with dim=None.

### LINKED ISSUE #121725: torch.logsumexp throws RuntimeError instead of reducing all dimensions when dim=None
### 🐛 Describe the bug

The [documentation ](https://pytorch.org/docs/stable/generated/torch.logsumexp.html#torch-logsumexp) of the dim argument of torch.logsumexp says: 

> dim ([int](https://docs.python.org/3/library/functions.html#int) or [tuple](https://docs.python.org/3/library/stdtypes.html#tuple) of ints, optional) – the dimension or dimensions to reduce. If None, all dimensions are reduced. 

However, when setting `dim=None`, which according to the documentation should reduce across all dimensions, the function throws a `RuntimeError` instead:

```python
import torch
print(torch.__version__) # 2.2.1+cu121
print(torch.logsumexp(torch.randn(3, 3), dim=(0, 1))) # tensor(2.4638)
print(torch.logsumexp(torch.randn(3, 3), dim=None))
```
Raises:
```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-3-ce5028272961>](https://localhost:8080/#) in <cell line: 4>()
      2 print(torch.__version__) #
      3 print(torch.logsumexp(torch.randn(3, 3), dim=(0, 1))) #
----> 4 print(torch.logsumexp(torch.randn(3, 3), dim=None)) #

RuntimeError: Please look up dimensions by name, got: name = None.
```

### Versions

Collecting environment information...
PyTorch version: 2.2.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.9
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.1.58+-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             2
On-line CPU(s) list:                0,1
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) CPU @ 2.20GHz
CPU family:                         6
Model:                              79
Thread(s) per core:                 2
Core(s) per socket:                 1
Socket(s):                          1
Stepping:                           0
BogoMIPS:                           4399.99
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          32 KiB (1 instance)
L1i cache:                          32 KiB (1 instance)
L2 cache:                           256 KiB (1 instance)
L3 cache:                           55 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0,1
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable; SMT Host state unknown
Vulnerability Meltdown:             Vulnerable
Vulnerability Mmio stale data:      Vulnerable
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:           Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] torch==2.2.1
[pip3] torchaudio==2.1.0+cu121
[pip3] torchdata==0.7.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.16.0
[pip3] torchvision==0.16.0+cu121
[pip3] triton==2.2.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by KungYork**
Fixes #121725

### CODE REVIEW

**Comment by Jonahmoon** [comment]
LG2M

**Comment by KungYork** [comment]
Seems that torch dynamo test is not related to this patch.

**Review by albanD — COMMENTED** [review]
I'll let @soulitzer handle this one but this looks BC-breaking?

**Comment by KungYork** [comment]
I'm not quite sure what 'BC breaking' refers to specifically. Could you provide more details? Thanks

**Comment by soulitzer** [comment]
> I'm not quite sure what 'BC breaking' refers to specifically. Could you provide more details? Thanks

'bc-breaking' means breaking backward compatibility, i.e., someone relying old behavior before this PR is landed will have their code broken

In this case, I'm not completely sure that it is bc-breaking since passing in `None` does not seem to be a valid input for named tensor logic either.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -1443,7 +1443,7 @@ Tensor nanmean(
 static Tensor& logsumexp_out_impl(Tensor& result, const Tensor& self, IntArrayRef dims, bool keepdim) {
   // can't take max of empty tensor
   if (self.numel() != 0) {
-    auto maxes = at::amax(self, dims, true);
+    auto maxes = at::amax(self, dims, dims.size()>0);
```
Comment (soulitzer) [inline aten/src/ATen/native/ReduceOps.cpp:1446]: nit: ```suggestion
    auto maxes = at::amax(self, dims, dims.size() > 0);
```

**Review by soulitzer — COMMENTED** [review]
Thanks for the fix! Could you add/update tests?

**Comment by KungYork** [comment]
Sure, I will add one.

**Pull request closed without merging**